tests/integration/record-arrays/adapter-populated-record-array-test.js

```diff
@@ -0,0 +1,278 @@
+  let recordArray = store.recordArrayManager
+    .createAdapterPopulatedRecordArray(store.modelFor('person'), null);
+
+  let payload = {
+    data: [
+      {
+        type: 'person',
+        id: '1',
+        attributes: {
+          name: 'Scumbag Dale'
+        }
+      },
+      {
+        type: 'person',
+        id: '2',
+        attributes: {
+          name: 'Scumbag Katz'
+        }
+      },
+      {
+        type: 'person',
+        id: '3',
+        attributes: {
+          name: 'Scumbag Bryn'
+        }
+      }
+    ]
+  };
+
+  run(() => {
+    recordArray._setInternalModels(store._push(payload), payload);
+  });
+
+  assert.equal(recordArray.get('length'), 3, "expected recordArray to contain exactly 3 records");
+
+  run(() => recordArray.get('firstObject').destroyRecord());
+
+  assert.equal(recordArray.get('length'), 2, "expected recordArray to contain exactly 2 records");
+});
+
+test('stores the metadata off the payload', function(assert) {
+  let recordArray = store.recordArrayManager
+    .createAdapterPopulatedRecordArray(store.modelFor('person'), null);
+
+  let payload = {
+    data: [
+      {
+        type: 'person',
+        id: '1',
+        attributes: {
+          name: 'Scumbag Dale'
+        }
+      },
+      {
+        type: 'person',
+        id: '2',
+        attributes: {
+          name: 'Scumbag Katz'
+        }
+      },
+      {
+        type: 'person',
+        id: '3',
+        attributes: {
+          name: 'Scumbag Bryn'
+        }
+      }
+    ],
+    meta: {
+      foo: 'bar'
+    }
+  };
+
+  run(() => {
+    recordArray._setInternalModels(store._push(payload), payload);
+  });
+
+  assert.equal(recordArray.get('meta.foo'), 'bar', 'expected meta.foo to be bar from payload');
+});
+
+test('stores the links off the payload', function(assert) {
+  let recordArray = store.recordArrayManager
+      .createAdapterPopulatedRecordArray(store.modelFor('person'), null);
+
+  let payload = {
+    data: [
+      {
+        type: 'person',
+        id: '1',
+        attributes: {
+          name: 'Scumbag Dale'
+        }
+      },
+      {
+        type: 'person',
+        id: '2',
+        attributes: {
+          name: 'Scumbag Katz'
+        }
+      },
+      {
+        type: 'person',
+        id: '3',
+        attributes: {
+          name: 'Scumbag Bryn'
+        }
+      }
+    ],
+    links: {
+      first: '/foo?page=1'
+    }
+  };
+
+  run(() => {
+    recordArray._setInternalModels(store._push(payload), payload);
+  });
+
+  assert.equal(recordArray.get('links.first'), '/foo?page=1', 'expected links.first to be "/foo?page=1" from payload');
+});
+
+test('recordArray.replace() throws error', function(assert) {
+  let recordArray = store.recordArrayManager
+    .createAdapterPopulatedRecordArray(Person, null);
+
+  assert.throws(() => {
+    recordArray.replace();
+  }, Error('The result of a server query (on (subclass of DS.Model)) is immutable.'), 'throws error');
```

Comment:
We're overwriting `toString()` in the `Person` above. Not sure why it's used here though.